src/main/scala/is/hail/utils/IntervalTree.scala

```diff
@@ -10,117 +10,128 @@ import scala.collection.mutable
 import scala.language.implicitConversions
 import scala.reflect.ClassTag
 
+case class IntervalEndpoint(point: Any, sign: Int) extends Serializable {
+  require(-1 <= sign && sign <= 1)
+}
+
+/**
+  * 'Interval' has an implicit precondition that there exists a Hail type
+  * 't: Type' such that either
+  * - 't: TBaseStruct', and 't.relaxedTypeCheck(left)', 't.relaxedTypeCheck(right),
+  *   and 't.ordering.intervalEndpointOrdering.lt(left, right)', or
+  * - 't.typeCheck(left)', 't.typeCheck(right)', and 't.ordering.lt(left, right)'
```

Comment:
I think that this block kind of means to me that "the interval has to have compatible points and can't be empty" and maybe that's worth stating somewhere in a more explicit but less precise way?